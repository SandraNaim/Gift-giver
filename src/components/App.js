import React, { Component } from "react";
import { Button } from "react-bootstrap";

class App extends Component {
  constructor(){
    super();
    this.state = {
      gifts: []
    }
  }

  addGifts = () => {
    const { gifts } = this.state
    const ids = gifts.map((gift) => gift.id)
    const max_id = ids.length > 0 ? Math.max(...ids) : 0
    gifts.push({id: max_id+1})
    this.setState({ gifts })
  }

  render(){
    return(
      <div>
        <h2>Gift Givers</h2>
        <div data-test="gift-list">
          {
            this.state.gifts.map((gift) => {
              return<div key={gift.id}></div>
            })
          }
        </div>
        <Button data-test="add-btn" onClick={this.addGifts}>Add gifts</Button>
      </div>
    )
  }
}

export default App;
