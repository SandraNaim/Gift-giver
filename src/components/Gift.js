import React, {Component} from "react";
import { Form, FormGroup, FormControl, FormLabel, Button } from "react-bootstrap";

class Gift extends Component {
    constructor(){
        super();
        this.state = {
            person: '',
            present: ''
        }
    }

    render(){
        return(
            <div className="gift">
                <Form>
                    <FormGroup>
                        <FormLabel>Person</FormLabel>
                        <FormControl 
                            data-test="person-input"
                            onChange={(evt) => this.setState({ person: evt.target.value })}
                        />
                    </FormGroup>
                    <FormGroup>
                        <FormLabel>Present</FormLabel>
                        <FormControl 
                            data-test="present-input"
                            onChange={(evt) => this.setState({ present: evt.target.value })}
                        />
                    </FormGroup>
                </Form>
                <Button
                    data-test="remove-button"
                    onClick={() => this.props.removeGift(this.props.gift.id)}
                >
                    Remove Gift
                </Button>
            </div>
        )
    }
}

export default Gift;