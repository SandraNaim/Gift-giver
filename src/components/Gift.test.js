import React from "react";
import { shallow } from "enzyme";
import Gift from './Gift';
import App from "./App";

describe('Gift', () => {
    const mockRemoveGift = jest.fn();
    const id = 1;
    const props = {gift: {id}, removeGift: mockRemoveGift}
    const gift = shallow(<Gift {...props} />);

    it('renders correctly', () => {
        expect(gift).toMatchSnapshot();
    });

    it('initialize empty string for the both states: person and present', () => {
        expect(gift.state()).toEqual({person: '', present: ''})
    });

    describe('while typing the person input', () => {
        const person = 'Uncle';
        beforeEach(() => {
            gift.find("[data-test='person-input']").simulate('change', { target: { value: person } })
        });

        it('update person state', () => {
            expect(gift.state().person).toEqual(person)
        });
    });

    describe('while typing the present input', () => {
        const present = 'bunny';
        beforeEach(() => {
            gift.find("[data-test='present-input']").simulate('change', { target: { value: present } })
        });

        it('update present state', () => {
            expect(gift.state().present).toEqual(present)
        });
    });

    describe('when clicking the `remove btn`', () => {
        beforeEach(() => {
            gift.find("[data-test='remove-button']").simulate('click');
        })

        test('remove gift', () => {
            expect(mockRemoveGift).toHaveBeenCalledWith(id)
        })
    })

})