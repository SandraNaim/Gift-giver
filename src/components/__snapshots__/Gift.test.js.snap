// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gift renders correctly 1`] = `
<div>
  <Form>
    <FormGroup>
      <FormLabel
        column={false}
        visuallyHidden={false}
      >
        Person Name
      </FormLabel>
      <FormControl
        data-test="person-input"
        onChange={[Function]}
      />
    </FormGroup>
    <FormGroup>
      <FormLabel
        column={false}
        visuallyHidden={false}
      >
        Present
      </FormLabel>
      <FormControl
        data-test="present-input"
        onChange={[Function]}
      />
    </FormGroup>
  </Form>
  <Button
    active={false}
    data-test="remove-button"
    disabled={false}
    onClick={[Function]}
    variant="primary"
  >
    Remove Gift
  </Button>
</div>
`;
